src/resolve/remote.rs

```diff
@@ -80,45 +82,53 @@ impl Remote {
 
     fn content_type(&self, response: &Response) -> Option<ContentType> {
         // Get a value content type header if any.
-        let header = response.headers().get(header::CONTENT_TYPE);
-        let header = header.and_then(|value| value.to_str().ok());
-
-        match header {
-            Some("text/markdown") => Some(ContentType::Markdown),
-            Some("image/png") | Some("image/jpeg") => Some(ContentType::Image),
-            Some("application/pdf") => Some(ContentType::Pdf),
+        let header = response.headers().get(header::CONTENT_TYPE)
+            .and_then(|raw| raw.to_str().ok())
+            .and_then(|string| string.parse::<mime::Mime>().ok())?;
+
+        match (header.type_(), header.subtype()) {
+            (mime::TEXT, sub) if sub == "markdown" => Some(ContentType::Markdown),
+            (mime::IMAGE, mime::PNG) | (mime::IMAGE, mime::JPEG) => Some(ContentType::Image),
+            (mime::APPLICATION, sub) if sub == "pdf" => Some(ContentType::Pdf),
+            // Let the file extension logic take over.
             _ => None,
         }
     }
 
+    /// Injectively map urls to paths in the download directory.
     fn target_path(&self, url: &Url) -> PathBuf {
         let mut target = self.temp.clone();
 
         // http(s) domains must contain a hostname
         target.push(url.host_str().unwrap());
 
-        // http(s) must not be cannot-be-base
-        //
-        // Also, '+' is a reserved character that can not appear unescaped.
-        let path = url.path().replace('/', "+");
+        // http(s) must not be cannot-be-base.  Also, '+' is a reserved character that can not
+        // appear unescaped and "+0" is a unique sequence to this replacement.
```

Comment:
I'm not too sure about this guarantee. The URL spec doesn't have this behaviour, any `+` should be treated literally. Only the HTML spec says that the URL should be application/x-www-form-urlencoded, which ensures the behaviour you're stating. See [this stackoverflow post](https://stackoverflow.com/questions/1005676/urls-and-plus-signs).

Reply:
Good point, `+` is only a reserved character to be used as a sub-delimiter. I misread the spec.

Reply:
Why don't you just SHA3 the URL and keep the file ending if it exists? Not only will that get rid of all parent-child structures so you don't need to manually handle them, but also generate a unique hash for every possible URL (not considering collisions obviously).